### PR TITLE
persist: split out a read-only runtime

### PR DESCRIPTION
**Pull request by danhhz**
See individual commits for details.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/9574)

### CODE REVIEW

File: src/persist/src/indexed/mod.rs
Comment (danhhz) [inline src/persist/src/indexed/mod.rs:1251]: these are all clunky but straightforward enough that it seemed okay to be. the only one I made a test wrapper for with the old signature was Register because it has to turn a bunch of things into owned strings, but curious if you think even that is worth it

File: src/persist/src/indexed/mod.rs
Comment (danhhz) [inline src/persist/src/indexed/mod.rs:50]: I vaguely recall some uneasy about how Register works but I've forgotten the details (you wanted it to take in both the string and id maybe?). now that we have this, I wonder if we could change the create_or_load protocol to fix that

File: src/persist/src/indexed/cache.rs
Comment (danhhz) [inline src/persist/src/indexed/cache.rs:65]: I'll reorder all these so there's one block when I do this for real

File: src/persist/src/indexed/mod.rs
Comment (danhhz) [inline src/persist/src/indexed/mod.rs:196]: do we think it's worth both of these sharing an Indexed? for example, it seems wrong for the read one to even have maintainer and pending. OTOH maintainer is also going away in the "full" version. since the read only commands mostly just call into state, the only code they really share is registering listeners and (soon) sending them things

File: src/persist/src/indexed/mod.rs
Comment (danhhz) [inline src/persist/src/indexed/mod.rs:70]: imagine a Replay or Tail or whatever command here that takes a `Vec<ListenEvent<Vec<u8>, Vec<u8>>`

File: src/persist/src/indexed/runtime.rs
Comment (danhhz) [inline src/persist/src/indexed/runtime.rs:1062]: no writing to blob means we don't have to batch these :)

File: src/persist/src/indexed/runtime.rs
Comment (danhhz) [inline src/persist/src/indexed/runtime.rs:270]: gonna move all these to be methods on the various enums

File: src/persist/src/indexed/runtime.rs
Comment (danhhz) [inline src/persist/src/indexed/runtime.rs:158]: we _could_ make this and `RuntimeImpl` share code, but these two are the ones where I really feel like a little duplication makes it all much easier to follow. wdyt?

File: src/persist/src/indexed/runtime.rs
Comment (danhhz) [inline src/persist/src/indexed/runtime.rs:228]: this might also later get a thread/task for dealing with `tail`ing the blob's META and turning it into ListenEvents

File: src/persist/src/indexed/runtime.rs
Comment (danhhz) [inline src/persist/src/indexed/runtime.rs:940]: I don't like this one, it's nearly an exact duplicate of `send`, but don't see anything nicer to do here. one option I suppose is to push Stop/Close into CmdRead and make Indexed handle it, but then the RuntimeImpl loops will have to sniff this out so it can flush out `step`, which is unsatisfying. I'm not convinced that there isn't something obviously much nicer to do here

**Comment by ruchirK** [comment]
I'll review this in the afternoon.

**Comment by ruchirK** [comment]
I'm really sorry I did not. Will look at this first thing after lunch (for realz this time 😅 )

File: src/persist/src/indexed/mod.rs
Comment (ruchirK) [reply src/persist/src/indexed/mod.rs:432]: I like apply. I dislike cmd.

File: src/persist/src/indexed/runtime.rs
Comment (ruchirK) [inline src/persist/src/indexed/runtime.rs:46]: I maybe have a mild preference to calling this variant `IndexedCmd`?

File: src/persist/src/indexed/mod.rs
Comment (ruchirK) [reply src/persist/src/indexed/mod.rs:70]: do you imagine a world where we have that _and_ Listen or do you think that would replace Listen?

File: src/persist/src/indexed/mod.rs
Comment (ruchirK) [reply src/persist/src/indexed/mod.rs:196]: both of these being Blob and BlobRead, and the alternative being to split out `IndexedRead<L: Log, B: BlobRead>` right? I think since the time this question was posed we have moved towards `IndexedRead`. Let me know if I'm missing something

File: src/persist/src/indexed/runtime.rs
Comment (ruchirK) [inline src/persist/src/indexed/runtime.rs:911]: I'm skimming so I may have missed something obvious but why is this variant called Full?

File: src/persist/src/indexed/runtime.rs
Comment (ruchirK) [reply src/persist/src/indexed/runtime.rs:160]: agree. we can also live with the duplication for a while and then decide to consolidate later?

**Review by ruchirK — COMMENTED** [review]
Did a quick first pass. Still reading through this. Going to switch to another task for a bit and will come back to this - definitely will be done reviewing by tonight.

File: src/persist/src/indexed/runtime.rs
Comment (danhhz) [reply src/persist/src/indexed/runtime.rs:911]: ha! this was from you, though I may have misremembered the term you used. the opposite of Read isn't really Write here, it's Read+Write. in most places we can get away with Foo and FooRead variants (the only real exception previous to this is open_exclusive vs open_read, which work because exclusive has its own meaning at the level of the storage trait). definitely open to suggestions

File: src/persist/src/indexed/runtime.rs
Comment (danhhz) [reply src/persist/src/indexed/runtime.rs:160]: nope. what my comment here is trying to get at is that DRYing them is one of those cases where it ends up being _less_ maintainable and I've intentionally kept them separate

File: src/persist/src/indexed/runtime.rs
Comment (danhhz) [reply src/persist/src/indexed/runtime.rs:46]: yeah that wfm

File: src/persist/src/indexed/mod.rs
Comment (danhhz) [reply src/persist/src/indexed/mod.rs:196]: yeah wrote this comment before we had that convo. that's the preferred plan at this point. only thing is this refactor is complicated enough that until i actually prototype it, i'll stay skeptical

File: src/persist/src/indexed/mod.rs
Comment (danhhz) [reply src/persist/src/indexed/mod.rs:70]: they're basically opposites, so we'll need them both

Comment (ruchirK) [reply src/persist/src/indexed/mod.rs:70]: is the distinction tail/replay sends from one (writer) persist runtime -> another (reader) persist runtime in a different process, whereas listen is the way a persist runtime sends updates to some other thread / component that's not part of persist running on the same process?

I dunno they definitely seem different but it seems to me like they're more similar than not - like it seems like `Tail` is like `Listen`, but over the network and to different persist runtimes -- is that a fair way of looking at it?

File: src/persist/src/indexed/mod.rs
Comment (ruchirK) [reply src/persist/src/indexed/mod.rs:1251]: I think these are fine it seems verbose but easily fits in one line and is straightforward like you said. I don't have strong feelings on the register bit -- I would maybe advocate for a higher level test helper like one that creates an index, registers a stream named "0", and fills it with some starting data, and returns the index and read/write handles or something? I feel like a lot of the tests do at least the first 2 things, and some do all 3?

File: src/persist/src/indexed/runtime.rs
Comment (ruchirK) [reply src/persist/src/indexed/runtime.rs:911]: ok it makes a lot more sense once the second variant is added in :D

File: src/persist/src/indexed/mod.rs
Comment (ruchirK) [reply src/persist/src/indexed/mod.rs:247]: this makes sense now.

File: src/persist/src/indexed/runtime.rs
Comment (ruchirK) [inline src/persist/src/indexed/runtime.rs:190]: nit: lets name this like `persist:runtime:read` or something like that?

**Review by ruchirK — COMMENTED** [review]
I did a fairly close read and everything seems to make sense! I think the only thing I'll need another pass on is the last two commits and specifically their changes to `runtime.rs` because the relationship between all the structs got a little bit complex. Overall though I think this is ready to be cleaned up!

**Comment by danhhz** [comment]
FYI I tried pulling out IndexedRead and played with it for a bit, but I couldn't manage to structure it in a way that made any sort of sense. Not to say it's impossible but we may have to punt on that for the initial version of this work

**Comment by danhhz** [comment]
Okay I still owe some more commentary in the commit messages and I'm sure there's some rustdoc that needs updating, but this is ready for a first pass!

**Comment by danhhz** [comment]
Oh I did the bad thing and github stole my ability to respond to comments directly :(

> nit: lets name this like persist:runtime:read or something like that?

Done. "persist:runtime:read" sounds child-parent-like to me, so went with "persist:runtime-read" since that sounded more sibling-like to me and IMO that's a closer characterization of the relationship between the two.

> I think these are fine it seems verbose but easily fits in one line and is straightforward like you said. I don't have strong feelings on the register bit -- I would maybe advocate for a higher level test helper like one that creates an index, registers a stream named "0", and fills it with some starting data, and returns the index and read/write handles or something? I feel like a lot of the tests do at least the first 2 things, and some do all 3?

Let's punt the helper for another time. This PR is already big and it doesn't seem like I've made things much worse

> the debug assert seems duplicated with the validate_pending_empty() calls in all of the match variants - I think that's just a consequence of this being a wip / prototype but wanted to double check that I wasn't missing something more fundamental

Sure, done. I preferred the explictness of the old way, but sounds like it made you think you were missing something

> I maybe have a mild preference to calling this variant IndexedCmd?

Done

> very very mild preference to just call this Read?

Done

> agree. we can also live with the duplication for a while and then decide to consolidate later?

I think this is one of those cases where DRY would make things _less_ maintainable

> both of these being Blob and BlobRead, and the alternative being to split out IndexedRead<L: Log, B: BlobRead> right? I think since the time this question was posed we have moved towards IndexedRead. Let me know if I'm missing something

I mentioned this elsewhere but I poked at this a bit and couldn't get anything that was even remotely understandable. I think there is something here but I don't see it yet

> I like apply. I dislike cmd.

Done

**Comment by danhhz** [comment]
Hey @ruchirK @aljoscha could y'all take a look at this tomorrow please? I want to stay ahead of any storage prototyping we might start doing and read-only persist will be a key part of that (plus this is gonna be a pain to rebase)

**Comment by ruchirK** [comment]
yes! sorry I saw your message go by but this wasn't in my "needs review" so it had slipped through the cracks. I will look at this either later tonight or first thing tomorrow

**Comment by aljoscha** [comment]
Reviewing now! I was going to ask how pressing this was, since I saw recent activity.

File: src/persist/src/indexed/mod.rs
Comment (aljoscha) [inline src/persist/src/indexed/mod.rs:493]: nit: could we still have internal methods for the individual `Cmd`s. As it is, the method is a bit unwieldy. It might be that you wanted to ensure that no one can by accident think they should call these methods instead of going through `apply`, if so, then please ignore.

File: src/persist/src/client.rs
Comment (aljoscha) [inline src/persist/src/client.rs:103]: nit: I prefer `Arc::clone()`, because it makes it more obvious that this is not a "real" clone. But: we could enable a lint for this and we don't, so ... :man_shrugging:

File: src/persist/src/client.rs
Comment (aljoscha) [inline src/persist/src/client.rs:224]: :clap:

File: src/persist/src/client.rs
Comment (aljoscha) [inline src/persist/src/client.rs:309]: nit: it looks like we have editor settings fighting here

File: src/persist/src/operators/source.rs
Comment (aljoscha) [inline src/persist/src/operators/source.rs:369]: :raised_hands:

File: src/persist/src/runtime.rs
Comment (aljoscha) [inline src/persist/src/runtime.rs:164]: nit: I'm guessing `pool` comes from when this was a thread pool and we didn't yet use an async runtime? Might be good to change it to `runtime` or `async_runtime` now? Same for `start`, above.

File: src/persist/src/runtime.rs
Comment (aljoscha) [inline src/persist/src/runtime.rs:160]: sorry if this is a stupid question, but: would this actually work as it is now? Could you create a read-only for a shared storage location from a separate process. This still takes a blob, which when created would clash with the lock of the "real" runtime? And this just has a regular blob/maintainer/indexed, which would theoretically be able to modify the original data.

I guess I'm just wondering about the state of this. Totally fine if the intend is to just sort out the client side. And I could be wrong above. :sweat_smile:

**Review by aljoscha — COMMENTED** [review]
The code looks good! I had some nits and a question about the state/readiness of this change, if one wanted to have a read-only runtime in a separate process. Driven by me not understanding the whole thing, mostly.

**Review by danhhz — COMMENTED** [review]
responded to a couple of your comments, will follow up with code changes

File: src/persist/src/client.rs
Comment (aljoscha) [reply src/persist/src/client.rs:309]: I'd be all for auto-formatting this as well!

File: src/persist/src/runtime.rs
Comment (aljoscha) [reply src/persist/src/runtime.rs:164]: No need to change, that reasoning also makes sense to me.

File: src/persist/src/client.rs
Comment (ruchirK) [reply src/persist/src/client.rs:309]: lol I didn't know that - i'm happy to auto format this too!

File: src/persist/src/runtime.rs
Comment (aljoscha) [reply src/persist/src/runtime.rs:160]: oh yeah, I missed the `BlobRead` bound on this one. :see_no_evil:

File: src/persist/src/indexed/mod.rs
Comment (ruchirK) [reply src/persist/src/indexed/mod.rs:94]: gentle ping on this question @danhhz

File: src/persist/src/indexed/mod.rs
Comment (ruchirK) [inline src/persist/src/indexed/mod.rs:587]: nit: reasonable

File: src/persist/src/client.rs
Comment (ruchirK) [inline src/persist/src/client.rs:38]: nit/random thought: we can probably express the "previous" relation in terms of `SeqNos` for mutating commands right?

File: src/persist/src/indexed/mod.rs
Comment (ruchirK) [reply src/persist/src/indexed/mod.rs:53]: I think it was a pretty abstract concern.

basically

if l had a sequence of commands like this in a log

```
load ("foo" ...)
load ("bar" ...)
write(Id(X), <data>)
write(Id(Y), <data2>)
```

and I told you "apply this sequence of commands to an empty persist runtime, and then tell me the states of all the collections" you would need some way to know which `Id(_)` mapped to which external stream name / would have to have access to the internals of how `persist` manages that. It's a tiny bit unsatisfying because the rest of the commands are all "self-describing" enough that anyone could replay them back and know exactly what to do.

I don't think this is a practical concern now or in the forseeable future.

File: src/persist/src/operators/input.rs
Comment (ruchirK) [inline src/persist/src/operators/input.rs:106]: nit: did you want to keep that as-is before merging? no strong preference from me just curious

**Review by ruchirK — APPROVED** [review]
Everything LGTM just left some small questions! sorry for the delay in reviewing

File: src/persist/src/indexed/mod.rs
Comment (danhhz) [reply src/persist/src/indexed/mod.rs:955]: done

File: src/persist/src/client.rs
Comment (danhhz) [reply src/persist/src/client.rs:309]: reverted

File: src/persist/src/runtime.rs
Comment (danhhz) [reply src/persist/src/runtime.rs:164]: fixed in the other PR!

File: src/persist/src/runtime.rs
Comment (danhhz) [reply src/persist/src/runtime.rs:160]: :thumbsup:

File: src/persist/src/indexed/mod.rs
Comment (danhhz) [reply src/persist/src/indexed/mod.rs:493]: threw something on your calendar for monday

File: src/persist/src/indexed/mod.rs
Comment (danhhz) [reply src/persist/src/indexed/mod.rs:587]: done

File: src/persist/src/operators/input.rs
Comment (danhhz) [reply src/persist/src/operators/input.rs:106]: yeah. i played with this for a bit but didn't come up with a fix I liked. plus I'd like to just delete this whole module, there's no use for it in mz

File: src/persist/src/indexed/mod.rs
Comment (danhhz) [reply src/persist/src/indexed/mod.rs:53]: ah right. well I think this helps move in that direction! we could replace Register with something that takes both the name and Id, make Load return the next free Id if it doesn't find it, and the impl create_or_load using a combination of those two (with care to handle races of course)

File: src/persist/src/client.rs
Comment (danhhz) [reply src/persist/src/client.rs:38]: IMO this is more actionable for a user of persist. SeqNos describe what happened, but this is talking about how you use persist to get the ordering guarantees you want

**Review by danhhz — COMMENTED** [review]
okay modulo the indexed::apply convo I have scheduled with aljoscha on monday, I think this is ready to merge! let me know if I missed something!

> is the distinction tail/replay sends from one (writer) persist runtime -> another (reader) persist runtime in a different process, whereas listen is the way a persist runtime sends updates to some other thread / component that's not part of persist running on the same process?
> 
> I dunno they definitely seem different but it seems to me like they're more similar than not - like it seems like Tail is like Listen, but over the network and to different persist runtimes -- is that a fair way of looking at it?

oh sorry, we discussed this in person at the end of the team sync yesterday so I thought it was resolved. we decided to go with the meta approach, which replaces Tail. but since you asked, I'll explain it for posterity :)

Tail wasn't (past tense because Tail is no longer a thing) like Listen at all. it was the exact opposite. Listen is how you get ListenEvents _out_ of a persist runtime. Tail was _ingecting_ a ListenEvent _into_ the read-only persist runtime for it to pass on to it's own listeners. maybe you got that ListenEvent from a Listen. maybe you made a fake one by manually diffing two METAs

anyway forget I said any of that because we're not doing Tail. the meta thing is way better and we're doing that instead

File: src/persist/src/indexed/mod.rs
Comment (danhhz) [reply src/persist/src/indexed/mod.rs:493]: discussed this offline and decided this is okay as-is

**Comment by danhhz** [comment]
okay rebased, TFTR!

**Pull request merged**
